Introduce `node-bench-regression-guard` to Substrate's pipeline

rcny opened the pull request:
Resolves paritytech/ci_cd#11 via itself and [`node-bench-regression-guard`](https://github.com/paritytech/scripts/tree/master/dockerfiles/node-bench-regression-guard) tool. 

This PR makes following changes to the project's pipeline:
* adds relevant `node-bench-regression-guard` job
* changes logic of `cargo-check-benches` job
* adds `cargo-check-benches-merged` job, which does the same thing as `cargo-check-benches`, but with PR's source branch already merged into `master`
* adds `verify` stage. `needs` is [unable](https://gitlab.com/gitlab-org/gitlab/-/issues/30632) to calculate the dependency graph when referring to the jobs in the same stage
* adds relevant YAML anchors

### Notes

`git fetch` with plus sign is used here for `cargo-check-benches-merged` to forcefully change needed refspecs to the actual ones. Runner [checkouts the specific pipeline's commit](https://gitlab.com/gitlab-org/gitlab/-/issues/15409), so, basically, the repository ends in the detached HEAD state. Additionally, default Runner's `GIT_STRATEGY` is to use `fetch`, so different runners have theirs own vision about what a current local `master` is.

`allow_failure` is needed for `node-bench-regression-guard` right now. There won't be any available artifacts for `master`'s bench results until this PR is merged, so the job will fail for the PR's pipelines. This directive will be removed shortly (via a separate PR) when `master` will be populated with bench results artifacts.

### Then how can I be sure that it works?

I've forked `substrate` on GitLab and merged these proposed changes to `master` there. Check it by yourself:
* ["This" PR](https://gitlab.parity.io/rcny/substrate/-/merge_requests/1)
* [Example PR](https://gitlab.parity.io/rcny/substrate/-/merge_requests/2) after `node-bench-regression-guard` introduction to Substrate's pipeline. [Relevant comparison job](https://gitlab.parity.io/rcny/substrate/-/jobs/879347)

More to it, `node-bench-regression-guard` has its own [test coverage](https://github.com/paritytech/scripts/blob/master/dockerfiles/node-bench-regression-guard/run-tests.rb).

TriplEight commented on .gitlab-ci.yml:
```diff
@@ -23,6 +23,7 @@
 stages:
   - check
   - test
+  - verify
```
I think the whole new stage is a bit of an overkill for the only job. Even though it runs only ~9sec, it holds other jobs to start. Eventually, these seconds add up to minutes. Our stage names are already pretty much vague, esp with DAG here and there, so unless you plan to depend the following stage on the state of this one, just put it into `build`.

rcny replied:
Resolved [here](https://github.com/paritytech/substrate/pull/8519/commits/7ed876a3e2dd43398832311b1b80f8aa3a3d8f89).

TriplEight commented on .gitlab-ci.yml:
```diff
@@ -96,6 +97,18 @@ default:
+    - if: $CI_PIPELINE_SOURCE == "schedule"
+    - if: $CI_COMMIT_REF_NAME == "tags"
+    - if: $CI_COMMIT_REF_NAME =~ /^[0-9]+$/                         # PRs
+    - if: $CI_COMMIT_REF_NAME =~ /^v[0-9]+\.[0-9]+.*$/              # i.e. v1.0, v2.1rc1
```
_philosohically:_ since this is a perf guard and it should prevent the slow code getting merged, does it make sense to check it again on tags?

rcny replied:
Resolved [here](https://github.com/paritytech/substrate/pull/8519/commits/7ed876a3e2dd43398832311b1b80f8aa3a3d8f89).

TriplEight commented on .gitlab-ci.yml:
```diff
@@ -223,11 +254,35 @@ cargo-check-benches:
   <<:                              *docker-env
   <<:                              *test-refs-no-trigger
+  <<:                              *cargo-check-benches-script
+  <<:                              *collect-artifacts
```
`*collect-artifacts` names them like this `"${CI_JOB_NAME}_${CI_COMMIT_REF_NAME}"`. I think it's more future proof to add some git sha information, either in a dir name or in some metadata. Just in case someone debugs the regression, so it would be easier to compare the results from different commits.

rcny replied:
Resolved [here](https://github.com/paritytech/substrate/pull/8519/commits/7ed876a3e2dd43398832311b1b80f8aa3a3d8f89).

TriplEight commented on .gitlab-ci.yml:
```diff
@@ -124,6 +137,24 @@ default:
+  script:
+    - mkdir -p artifacts/benches/$CI_COMMIT_REF_NAME
+    - SKIP_WASM_BUILD=1 time cargo +nightly check --benches --all
+    - 'cargo run --release -p node-bench -- ::node::import::native::sr25519::transfer_keep_alive::paritydb::small --json
```
quotes here to avoid gitlab issues with colons?

rcny replied:
exactly

TriplEight reviewed (changes requested):
Looks good, just some nits

rcny commented:
@TriplEight, applied your suggestions:
* removed `verify` stage
* moved `node-bench-regression-guard` to `build` stage
* added `$CI_COMMIT_SHORT_SHA` to artifacts' directories
* changed `rules` job inclusion logic: only PRs are valid refs now
* moved some scripts to the relevant anchors. it resolves previous `before_script` override for `cargo-check-benches-merged` job. now it shows `rustup`/`cargo` information correctly

rcny commented:
Updated [comparison job](https://gitlab.parity.io/rcny/substrate/-/jobs/885785) from my fork.
